aspnetcore/blazor/javascript-interoperability/static-server-rendering.md

```diff
@@ -0,0 +1,141 @@
+---
+title: ASP.NET Core Blazor JavaScript with Blazor static Server rendering
+author: guardrex
+description: Learn how to use JavaScript in a Blazor Web App with static Server rendering.
+monikerRange: '>= aspnetcore-8.0'
+ms.author: riande
+ms.custom: mvc
+ms.date: 11/03/2023
+uid: blazor/js-interop/ssr
+---
+# ASP.NET Core Blazor JavaScript with Blazor static Server rendering
+
+<!-- UPDATE 9.0 Activate after release and INCLUDE is updated
+
+[!INCLUDE[](~/includes/not-latest-version.md)]
+
+-->
+
+This article explains how to load JavaScript (JS) in a Blazor Web App with static Server rendering and [enhanced navigation](xref:blazor/fundamentals/routing#enhanced-navigation-and-form-handling).
+
+Some applications may depend on JS to perform initialization tasks that are specific to each page. When using Blazor's enhanced navigation feature, which allows the user to avoid reloading the entire page, JS code may not re-execute as expected. Therefore, page-specific `<script>` elements may not be executed again as expected each time an enhanced page navigation occurs.
+
+To avoid this problem, we don't recommended relying on page-specific `<script>` elements outside of the app's common layout. Instead, scripts should register an [`afterWebStarted` JS initializer](xref:blazor/fundamentals/startup#javascript-initializers) to perform initialization logic and use an event listener (`blazor.addEventListener("enhancedload", callback)`) to listen for page updates caused by enhanced navigation.
+
+The following example demonstrates one way to configure JS code to run when a page is initially loaded or updated.
+
+In the Blazor Web App, add the following `EnhancedLoadScript` component.
+
+`Components/Pages/EnhancedLoadScript.razor`:
+
+```razor
+@page "/enhanced-load-script"
+@using BlazorPageScript
+
+<PageTitle>Enhanced Load Script Example</PageTitle>
+
+<PageScript Src="js/pages/home.js" />
+
+Welcome to my page.
+```
+
+In the Blazor Web App, add the following JS file (`home.js`):
+
+* The `onLoad` function is called when the page loads the first time.
+* The `onUpdate` function is called each time an enhanced load occurs on the page, plus once when the page is first loaded.
+* The `onDispose` function is called when the user leaves the page due to an enhanced navigation.
+
+The `home.js` file is placed in the `wwwroot` folder and not [collocated with the component](xref:blazor/js-interop/index#load-a-script-from-an-external-javascript-file-js-collocated-with-a-component) in order to make it reusable around the app from a common location.
```

Comment:
It could still be a collocated script - in fact it might be preferable that way. This example is meant to demonstrate page-specific initialization logic.

That said, it's an interesting idea to support the case where a script could be reusable between pages. The `<PageScript>` component could be changed so it works in a more generalized manner:

* It calls 'onLoad' when the script gets added to the page
* It calls 'onUpdate' when the script still exists on the page after an enhanced update
* It calls 'onDispose' when the script gets removed from the page after an enhanced update

So it works without regard to current URL. Doing it this way would mean the developer gets more control over the lifetime of the script.

For example, if you put a `<PageScript>` on the top-level of a page, then it the behavior is no different from the way the example is now. But if you put it in the app's layout (making it shared between pages), then it only runs "load" after a full page reload, and "update" when any enhanced page update occurs (including enhanced navigation).

Or, if you want to reuse the same module between pages, but get the "load" and "dispose" callbacks invoked on each page change, you could append a querystring to the end of the script so it's recognized as a different module:

`<PageScript Src="./page.js?counter" />`

I might go ahead and propose a new version of the implementation that has these characteristics.

Reply:
> It could still be a collocated script - in fact it might be preferable that way.

This was one of my *incorrect* guesses, due to you not colocating the script. It will be fixed up on my next commit.